File: ios/LinkFlowSDKModule.swift
import Foundation
import React

/// React Native bridge for LinkFlowSDK.
///
/// Wraps the canonical Swift SDK (vendored from linkflow-ios-sdk) and exposes
/// it to JavaScript as the `LinkFlowSDK` native module. Caches the most recent
/// attribution result locally so JS callers can fetch it synchronously via
/// `getAttributionResult` even after the originating callback has fired.
@objc(LinkFlowSDKModule)
public class LinkFlowSDKModule: RCTEventEmitter {

    private var hasListeners = false
    private var lastResult: AttributionResult?

    // MARK: - RCTEventEmitter overrides

    @objc public override static func requiresMainQueueSetup() -> Bool { false }

    public override func supportedEvents() -> [String]! {
        return ["LinkFlowAttribution", "LinkFlowDeepLink", "LinkFlowError"]
    }

    public override func startObserving() { hasListeners = true }
    public override func stopObserving()  { hasListeners = false }

    // MARK: - Exposed methods

    @objc(initialize:enableLogging:resolver:rejecter:)
    func initialize(_ apiBaseURL: String,
                    enableLogging: Bool,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
        let sdk = LinkFlowSDK.initialize(apiBaseURL: apiBaseURL, enableLogging: enableLogging)
        sdk.setAttributionCallback { [weak self] result in
            self?.handleResult(result)
        }
        resolve(nil)
    }

    @objc(handleAppLaunch:resolver:rejecter:)
    func handleAppLaunch(_ initialUrl: NSString?,
                         resolver resolve: @escaping RCTPromiseResolveBlock,
                         rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            let sdk = try LinkFlowSDK.shared()
            // Reattach callback (bridge may be recreated across reloads).
            sdk.setAttributionCallback { [weak self] result in
                self?.handleResult(result)
            }
            sdk.handleAppLaunch(userActivity: nil)
            // If a launch URL is provided, treat it as a custom-scheme deep link.
            if let urlString = initialUrl as String?,
               !urlString.isEmpty,
               let url = URL(string: urlString) {
                _ = sdk.handleCustomURL(url: url)
            }
            resolve(nil)
        } catch {
            reject("not_initialized", error.localizedDescription, error)
        }
    }

    @objc(handleDeepLink:resolver:rejecter:)
    func handleDeepLink(_ urlString: NSString,
                        resolver resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let url = URL(string: urlString as String) else {
            reject("invalid_url", "Invalid URL: \(urlString)", nil)
            return
        }
        do {
            let sdk = try LinkFlowSDK.shared()
            _ = sdk.handleCustomURL(url: url)
            // Also emit a JS-level deep-link event for consumers that want
            // raw URL notifications independent of attribution callbacks.
            if hasListeners {
                sendEvent(withName: "LinkFlowDeepLink", body: ["url": url.absoluteString])
            }
            resolve(nil)
        } catch {
            reject("not_initialized", error.localizedDescription, error)
        }
    }

    @objc(trackEvent:params:revenue:resolver:rejecter:)
    func trackEvent(_ eventName: NSString,
                    params: NSDictionary?,
                    revenue: NSNumber?,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            let sdk = try LinkFlowSDK.shared()
            sdk.trackEvent(
                eventName: eventName as String,
                params: params as? [String: Any],
                revenue: revenue
            )
            resolve(nil)
        } catch {
            reject("not_initialized", error.localizedDescription, error)
        }
    }

    @objc(getAttributionResult:rejecter:)
    func getAttributionResult(_ resolve: @escaping RCTPromiseResolveBlock,
                              rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let cached = lastResult else {
            resolve(nil)
            return
        }
        resolve(serialize(cached))
    }

    // MARK: - Internal

    private func handleResult(_ result: AttributionResult) {
        self.lastResult = result
        if let error = result.error {
            if hasListeners {
                sendEvent(withName: "LinkFlowError", body: ["message": "\(error)"])
            }
            return
        }
        if hasListeners {
            sendEvent(withName: "LinkFlowAttribution", body: serialize(result))
        }
    }

    private func serialize(_ result: AttributionResult) -> [String: Any] {
        var out: [String: Any] = [
            "attributed": result.attributed,
            "deepLinkParams": result.deepLinkParams,
            "campaignData": result.campaignData
        ]
        if let v = result.deepLinkValue { out["deepLinkValue"] = v }
        return out
    }
}

File: android/src/main/kotlin/com/linkflow/sdk/rn/LinkFlowSDKModule.kt
package com.linkflow.sdk.rn

import android.content.Intent
import android.net.Uri
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.linkflow.sdk.LinkFlowSDK

/**
 * React Native bridge for the canonical Android LinkFlow SDK.
 *
 * Wraps the vendored [LinkFlowSDK] and exposes it as the `LinkFlowSDK`
 * native module on the JS side. Caches the most recent attribution result
 * so JS callers can fetch it via [getAttributionResult] without waiting for
 * another callback.
 */
class LinkFlowSDKModule(
    private val reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext), LinkFlowSDK.AttributionCallback {

    private var lastResult: LinkFlowSDK.AttributionResult? = null

    override fun getName(): String = NAME

    @ReactMethod
    fun initialize(apiBaseURL: String, enableLogging: Boolean, promise: Promise) {
        try {
            val sdk = LinkFlowSDK.initialize(reactContext.applicationContext, apiBaseURL, enableLogging)
            sdk.setAttributionCallback(this)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("init_failed", e.message, e)
        }
    }

    @ReactMethod
    fun handleAppLaunch(initialUrl: String?, promise: Promise) {
        try {
            val sdk = LinkFlowSDK.getInstance()
            sdk.setAttributionCallback(this)

            // If JS supplied a launch URL, wrap it in a synthetic Intent so the
            // canonical SDK's first-launch attribution path can extract the
            // click_token query parameter.
            val intent: Intent? = if (!initialUrl.isNullOrEmpty()) {
                Intent(Intent.ACTION_VIEW, Uri.parse(initialUrl))
            } else {
                currentActivity?.intent
            }
            sdk.handleAppLaunch(intent)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("handle_app_launch_failed", e.message, e)
        }
    }

    @ReactMethod
    fun handleDeepLink(url: String, promise: Promise) {
        try {
            // The canonical SDK does not expose a public deep-link entry point
            // for already-running apps, so emit the event directly to JS.
            val map = Arguments.createMap()
            map.putString("url", url)
            emit(EVENT_DEEPLINK, map)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("handle_deeplink_failed", e.message, e)
        }
    }

    @ReactMethod
    fun trackEvent(eventName: String, params: ReadableMap?, revenue: Double?, promise: Promise) {
        try {
            val map = params?.toHashMap()
            LinkFlowSDK.getInstance().trackEvent(eventName, map, revenue)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("track_event_failed", e.message, e)
        }
    }

    @ReactMethod
    fun getAttributionResult(promise: Promise) {
        try {
            val r = lastResult
            if (r == null) {
                promise.resolve(null)
            } else {
                promise.resolve(serializeResult(r))
            }
        } catch (e: Exception) {
            promise.reject("get_result_failed", e.message, e)
        }
    }

    // RN's NativeEventEmitter expects these to exist on the native module.
    @ReactMethod fun addListener(eventName: String) { /* no-op */ }
    @ReactMethod fun removeListeners(count: Int) { /* no-op */ }

    // ----- AttributionCallback -----

    override fun onAttributionResolved(result: LinkFlowSDK.AttributionResult) {
        lastResult = result
        emit(EVENT_ATTRIBUTION, serializeResult(result))
    }

    override fun onAttributionError(error: Throwable) {
        val map = Arguments.createMap()
        map.putString("message", error.message ?: "unknown error")
        emit(EVENT_ERROR, map)
    }

    override fun onDeepLinkReceived(uri: Uri) {
        val map = Arguments.createMap()
        map.putString("url", uri.toString())
        emit(EVENT_DEEPLINK, map)
    }

    // ----- Helpers -----

    private fun emit(name: String, body: WritableMap) {
        if (!reactContext.hasActiveReactInstance()) return
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(name, body)
    }

    private fun serializeResult(r: LinkFlowSDK.AttributionResult): WritableMap {
        val map = Arguments.createMap()
        map.putBoolean("attributed", r.attributed)
        if (r.deepLinkValue != null) map.putString("deepLinkValue", r.deepLinkValue)
        map.putMap("deepLinkParams", mapToWritable(r.deepLinkParams))
        map.putMap("campaignData", mapToWritable(r.campaignData))
        return map
    }

    @Suppress("UNCHECKED_CAST")
    private fun mapToWritable(input: Map<String, Any?>): WritableMap {
        val out = Arguments.createMap()
        for ((k, v) in input) {
            when (v) {
                null -> out.putNull(k)
                is String -> out.putString(k, v)
                is Boolean -> out.putBoolean(k, v)
                is Int -> out.putInt(k, v)
                is Long -> out.putDouble(k, v.toDouble())
                is Double -> out.putDouble(k, v)
                is Float -> out.putDouble(k, v.toDouble())
                is Map<*, *> -> out.putMap(k, mapToWritable(v as Map<String, Any?>))
                else -> out.putString(k, v.toString())
            }
        }
        return out
    }

    companion object {
        const val NAME = "LinkFlowSDK"
        private const val EVENT_ATTRIBUTION = "LinkFlowAttribution"
        private const val EVENT_DEEPLINK = "LinkFlowDeepLink"
        private const val EVENT_ERROR = "LinkFlowError"
    }
}
